flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/BufferBuilder.java

```diff
@@ -86,11 +94,34 @@ public int append(ByteBuffer source) {
 		int available = getMaxCapacity() - positionMarker.getCached();
 		int toCopy = Math.min(needed, available);
 
+		// Each Data BufferBuilder starts with a 4-byte integer header
+		// Since length can only be 0 or positive numbers, the first bit of the integer
+		// is used to identify whether the first record is partial (1) or not (0)
+		// The remaining 31 bits stands for the length of the record remaining.
+		// The data written is not made visible to reader until {@link #commit()}, so the BufferBuilder
+		// ends either with a complete record or full buffer after append();
+		if (isEmptyBufferBuilder()) {
+			available = available - BUFFER_BUILDER_HEADER_SIZE;
+			toCopy = Math.min(needed, available);
```

Comment:
Can `available` be less than `BUFFER_BUILDER_HEADER_SIZE`?
Then `toCopy` can be negative.